src/storage/src/hummock/utils.rs

```diff
@@ -127,7 +113,39 @@ where
         let ord = user_key(&table.key_range.as_ref().unwrap().left).cmp(key.as_ref());
         ord == Ordering::Less || ord == Ordering::Equal
     })
-    .saturating_sub(1) // considering the boundary of 0
+}
+
+/// Prune overlapping SSTs that does not overlap with a specific key range or does not overlap with
+/// a specific table id. Returns the sst ids after pruning.
+pub fn prune_overlapping_ssts<'a, R, B>(
+    ssts: &'a [SstableInfo],
+    table_id: &'a TableId,
+    table_key_range: &'a R,
+) -> impl DoubleEndedIterator<Item = &'a SstableInfo>
+where
+    R: RangeBounds<TableKey<B>>,
+    B: AsRef<[u8]>,
+{
+    ssts.iter()
+        .filter(|info| filter_single_sst(info, table_id, table_key_range))
+}
+
+/// Prune non-overlapping SSTs that does not overlap with a specific key range or does not overlap
+/// with a specific table id. Returns the sst ids after pruning.
+pub fn prune_nonoverlapping_ssts<'a>(
+    ssts: &'a [SstableInfo],
+    encoded_user_key_range: &'a (Bound<Vec<u8>>, Bound<Vec<u8>>),
```

Comment:
Instead of encoding table_key_range in `prune_nonoverlapping_ssts`, it is to reduce the encoding of each call, right?

Reply:
This is purely a refactoring so the previous logic is preserved: use encoded_user_key_range to prune non-overlapping SSTs and use table_key_range to prune overlapping SSTs. I want to unify these two also but I think it is better to do it in a separate PR.